### PR TITLE
Store Model class reference on an instance of index class. Fixes #992

### LINKED ISSUE #992: Query on GSI not respecting polymorphism filter condition
Hello everyone! Great library.

The [polymorphism documentation](https://pynamodb.readthedocs.io/en/latest/polymorphism.html) states: "Query and scan operations transparently add a filter condition to ensure that only items with a matching discriminator value are returned."

However, when running a query on a GSI using a child class the filter condition does not seem to be correctly added as I still receive results matching another child class.

Setup:
```
class MyGSI(GlobalSecondaryIndex):
    class Meta:
        index_name = 'foo-index'
        read_capacity_units = 2
        write_capacity_units = 1
        projection = AllProjection()

    modelId = UnicodeAttribute(hash_key=True)

class ParentModel(Model):
    class Meta:
        table_name = 'polymorphic_table'
    id = UnicodeAttribute(hash_key=True)
    modelId = UnicodeAttribute()
    my_gsi = MyGSI()
    cls = DiscriminatorAttribute()

class FooModel(ParentModel, discriminator='Foo'):
    foo = UnicodeAttribute()

class BarModel(ParentModel, discriminator='Bar'):
    bar = UnicodeAttribute()
```

Then

```
FooModel.my_gsi.query(...)
```

will also return results from `BarModel`.

Is there a mistake in my setup or is this a bug?

### PR DESCRIPTION
**Pull request by jpinner-lyft**
This is a breaking change to the index API. The count, query, and scan methods are now instance methods.

### CODE REVIEW

**Comment by jpinner-lyft** [comment]
@ikonst @garrettheel so this "works" but breaks backwards compatibility a little bit and is certainly not elegant

**Comment by jpinner-lyft** [comment]
A lot of the complexity here comes from storing the Model class reference on the index Meta class itself.
If we instead stored the reference on the instance of the index in the Model class that simplifies things slightly.
That would be a slightly larger breaking change though.

**Comment by ikonst** [comment]
> That would be a slightly larger breaking change though.

In what sense? Do you think the fact that we use the `Index.Meta.model` is part of the contract somehow?

**Comment by jpinner-lyft** [comment]
@ikonst right, removing the classmethod annotation and then changing how an end user could set the model reference

**Comment by ikonst** [comment]
For maximum compatibility, maybe we can do some atrocity such as a method that's both class and instance:
```
class Foo:
    def bar(self=None):
        self = self or Foo
        ...
```

In the "classmethod" path, it'll keep using the old Meta.model. In the "instance" path, it'll use a newly-introduced Index instance field.

Maybe that's too crazy, but just throwing it out there.

**Comment by jpinner-lyft** [comment]
@ikonst I implemented your suggestion, let me know what you think.

I had to replace `Index.Meta.model` with `Index._model` so that I could override the lookup on an instance or class basis.

Then I wrote a descriptor `classorinstancemethod` which passes in either the class or the instance as the first argument.

Finally during Model class instantiation, I store a reference to the Model on the index class if it hasn't been set to make sure `Index.query` and the like keep working while creating a local class to store a reference to the current model class so that subclassing/polymorphism still works.

File: pynamodb/indexes.py [hunk 1]
```diff
@@ -48,16 +48,25 @@ class Index(Generic[_M], metaclass=IndexMeta):
...
+    _model: _M
 
     def __init__(self) -> None:
         if self.Meta is None:
             raise ValueError("Indexes require a Meta class for settings")
         if not hasattr(self.Meta, "projection"):
             raise ValueError("No projection defined, define a projection for this class")
 
-    @classmethod
+    class classorinstancemethod:
```
Comment (ikonst) [inline pynamodb/indexes.py:59]: nit - if we're keeping it in this file, I'd move it to module scope and prefix with underscore. But maybe move it to pynamodb/_compat.py?

File: pynamodb/indexes.py [hunk 2]
```diff
@@ -48,16 +48,25 @@ class Index(Generic[_M], metaclass=IndexMeta):
...
+        def __init__(self, func):
+            self.func = func
+        def __get__(self, instance, owner):
+            def _method(*args, **kwargs):
+                return self.func(instance or owner, *args, **kwargs)
+            return _method
```
Comment (ikonst) [inline pynamodb/indexes.py:65]:
```suggestion
            return functools.partial(self.func, instance or owner)
```

File: pynamodb/indexes.py [hunk 3]
```diff
@@ -48,16 +48,25 @@ class Index(Generic[_M], metaclass=IndexMeta):
...
+
+    @classorinstancemethod
     def count(
-        cls,
+        obj,
```
Comment (ikonst) [inline pynamodb/indexes.py:69]:
```suggestion
        obj: Union[Index, Type[Index]],
```
?

**Comment by garrettheel** [comment]
@jpinner-lyft we're looking at bumping major versions for #1003, do you want to get this in before that? Otherwise I'm fine with going for the breaking change here and making it available from v6 onwards

cc @ikonst

File: pynamodb/indexes.py [hunk 4]
```diff
@@ -43,14 +44,13 @@ def __init__(self) -> None:
             raise ValueError("No projection defined, define a projection for this class")
 
     def __set_name__(self, owner: Type[_M], name: str):
-        if not hasattr(self.Meta, "model"):
-            self.Meta.model = owner
+        if not hasattr(type(self), "_model"):
```
Comment (garrettheel) [inline pynamodb/indexes.py:47]: Do we still need this? Why wouldn't `_initialize_indexes()` set it?

Comment (jpinner-lyft) [reply pynamodb/indexes.py:47]: good catch -- i don't think we do now since we never need to access the variable on a class

**Review by garrettheel — APPROVED** [review]

**Pull request merged**